### PR TITLE
Simplify SparseMatrix setindex!

### LINKED ISSUE #27013: sparse `setindex!` with boolean indices broken
E.g.:
```julia
julia> using SparseArrays: spzeros

julia> a = spzeros(2,2)
2×2 SparseArrays.SparseMatrixCSC{Float64,Int64} with 0 stored entries

julia> a[[true,false],[true, false]] = a[[true,false],[true, false]]
ERROR: DimensionMismatch("")
Stacktrace:
 [1] setindex!(::SparseArrays.SparseMatrixCSC{Float64,Int64}, ::SparseArrays.SparseMatrixCSC{Float64,Int64}, ::Array{Bool,1}, ::Array{Bool,1}) at usr/share/julia/stdlib/v0.7/SparseArrays/src/sparsematrix.jl:2536
 [2] top-level scope

julia> a[:,[true, false]] = a[:,[true, false]]
ERROR: DimensionMismatch("")
Stacktrace:
 [1] setindex!(::SparseArrays.SparseMatrixCSC{Float64,Int64}, ::SparseArrays.SparseMatrixCSC{Float64,Int64}, ::UnitRange{Int64}, ::Array{Bool,1}) at usr/share/julia/stdlib/v0.7/SparseArrays/src/sparsematrix.jl:2536
 [2] setindex!(::SparseArrays.SparseMatrixCSC{Float64,Int64}, ::SparseArrays.SparseMatrixCSC{Float64,Int64}, ::Colon, ::Array{Bool,1}) at usr/share/julia/stdlib/v0.7/SparseArrays/src/sparsematrix.jl:2358
 [3] top-level scope

julia> a[1,[true, false]] = a[1,[true, false]]
ERROR: DimensionMismatch("parent has 1 elements, which is incompatible with size (1, 2)")
Stacktrace:
 [1] _throw_dmrs(::Int64, ::String, ::Tuple{Int64,Int64}) at ./reshapedarray.jl:157
 [2] _reshape at ./reshapedarray.jl:152 [inlined]
 [3] reshape at ./reshapedarray.jl:93 [inlined]
 [4] reshape at ./reshapedarray.jl:96 [inlined]
 [5] setindex!(::SparseArrays.SparseMatrixCSC{Float64,Int64}, ::SparseArrays.SparseVector{Float64,Int64}, ::Int64, ::Array{Bool,1}) at usr/share/julia/stdlib/v0.7/SparseArrays/src/sparsematrix.jl:2528
 [6] top-level scope
```
The assignments all work when using `a=zeros(2,2)` instead. 
Caused by #26347, so hopefully an easy fix for @mbauman.

### PR DESCRIPTION
**Pull request by mbauman**
Fix #27013, and add a whole slew of tests. `methods(setindex!, Tuple{SparseMatrixCSC, Vararg{Any}})` goes from a hodgepodge of 26 methods that were tough to reason about to just 6 methods: scalar, logical linear indexing, linear indexing with a vector, linear indexing with a matrix, generic nonscalar indexing, and a disambiguation method. This is much easier to reason about, and fixes a handful of bugs.

### CODE REVIEW

File: stdlib/SparseArrays/src/deprecated.jl [hunk 1]
```diff
@@ -223,13 +223,7 @@ import Base: asyncmap
...
 
 # PR 26347: implicit scalar broadcasting within setindex!
-@deprecate setindex!(A::SparseMatrixCSC, x::Number, i::Integer, J::AbstractVector{<:Integer}) (A[i, J] .= x; A)
-@deprecate setindex!(A::SparseMatrixCSC, x::Number, I::AbstractVector{<:Integer}, j::Integer) (A[I, j] .= x; A)
-@deprecate setindex!(A::SparseMatrixCSC, x, ::Colon)          fill!(A, x)
-@deprecate setindex!(A::SparseMatrixCSC, x, ::Colon, ::Colon) fill!(A, x)
-@deprecate setindex!(A::SparseMatrixCSC, x, ::Colon, j::Union{Integer, AbstractVector}) (A[:, j] .= x; A)
-@deprecate setindex!(A::SparseMatrixCSC, x, i::Union{Integer, AbstractVector}, ::Colon) (A[i, :] .= x; A)
-@deprecate setindex!(A::SparseMatrixCSC, x::Number, I::AbstractVector{<:Integer}, J::AbstractVector{<:Integer}) (A[I, J] .= x; A)
+@deprecate setindex!(A::SparseMatrixCSC{<:Any,<:Any}, x, i::Union{Integer, AbstractVector{<:Integer}, Colon}, j::Union{Integer, AbstractVector{<:Integer}, Colon}) (A[i, :] .= x; A)
```
Comment (martinholters) [inline stdlib/SparseArrays/src/deprecated.jl]: `A[i, j]`

**Comment by martinholters** [comment]
Superficially, LGTM, modulo the bug in the deprecation. Thanks for going after this so quickly!

File: stdlib/SparseArrays/src/deprecated.jl (hunk 1 above)
Comment (mbauman) [reply stdlib/SparseArrays/src/deprecated.jl]: Yikes, thank you!

File: stdlib/SparseArrays/src/sparsematrix.jl [hunk 2]
```diff
@@ -2520,21 +2515,15 @@ function _spsetnz_setindex!(A::SparseMatrixCSC{Tv}, x::Tv,
...
+_to_same_csc(::SparseMatrixCSC{Tv, Ti}, V::AbstractVector, I...) where {Tv,Ti} = convert(SparseMatrixCSC{Tv,Ti}, reshape(V, map(length, I)))
 
-# Nonscalar A[I,J] = B
-setindex!(A::SparseMatrixCSC{Tv,Ti}, B::SparseMatrixCSC{Tv,Ti}, I::Integer, J::Integer) where {Tv,Ti} =
-    setindex!(A, convert(Tv, I, J), I, J)
-function setindex!(A::SparseMatrixCSC{Tv,Ti}, B::SparseMatrixCSC{Tv,Ti}, I::Union{Integer, AbstractVector{<:Integer}}, J::Union{Integer, AbstractVector{<:Integer}}) where {Tv,Ti}
-    if size(B,1) != length(I) || size(B,2) != length(J)
-        throw(DimensionMismatch(""))
-    end
+setindex!(A::SparseMatrixCSC{Tv}, B::AbstractVecOrMat, I::Integer, J::Integer) where {Tv} = setindex!(A, convert(Tv, B), I, J)
```
Comment (Sacha0) [inline stdlib/SparseArrays/src/sparsematrix.jl]: The `convert(Tv, B)` call where `B::AbstractVecOrMat` confuses me; is `Tv` not usually scalar here?

Comment (martinholters) [reply stdlib/SparseArrays/src/sparsematrix.jl]: `I` and `J` are `Integer`s, so this sets one element in `A`. If `Tv` is indeed scalar here, that assignment should fail, which it does by the failing `convert` call.

**Comment by mbauman** [comment]
Failures were all #27041, but merging with so many ❌s gives me the heeby-jeebies.

File: stdlib/SparseArrays/src/sparsematrix.jl (hunk 2 above)
Comment (Sacha0) [reply stdlib/SparseArrays/src/sparsematrix.jl]: Thanks for the awesome explanation gentlemen! :)

**Pull request merged**